# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [32]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [33]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [34]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [35]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [36]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [37]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [38]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [39]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [40]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [41]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [42]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [43]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [44]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [45]:
unique_seasons = df['Season'].nunique()
print("Number of unique seasons in the dataset:", unique_seasons)

Number of unique seasons in the dataset: 32


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [46]:
import pandas as pd

# Assuming 'df' is my pandas DataFrame and 'Wteam' is the column with the winning team data

team_most_wins = df['Wteam'].value_counts().idxmax()
number_of_wins = df['Wteam'].value_counts().max()
print("Team with the most wins:", team_most_wins)

Team with the most wins: 1181


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [16]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [17]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [18]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [19]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [20]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [21]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [22]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [23]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [24]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [25]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [47]:
import pandas as pd

# Let's assume 'data' is my pandas DataFrame

# Sorting by 'season'

df_season_sorted = df.sort_values('Season')[['Season', 'Wteam', 'Wscore']].iloc[100:200]

# Sorting by 'winning team'

df_wteam_sorted = df.sort_values('Wteam')[['Season', 'Wteam', 'Wscore']].iloc[100:200]

# Sorting by 'winning score'

df_wscore_sorted = df.sort_values('Wscore')[['Season', 'Wteam', 'Wscore']].iloc[100:200]

# I want to display the results:
print(df_season_sorted.head())
print(df_wteam_sorted.head())
print(df_wscore_sorted.head())

      Season  Wteam  Wscore
2525    1985   1406      68
2526    1985   1410      83
2527    1985   1423      58
2528    1985   1429      86
2529    1985   1440      50
       Season  Wteam  Wscore
97792    2008   1102      71
97721    2008   1102      67
96770    2007   1102      67
96658    2007   1102      69
96454    2007   1102      60
        Season  Wteam  Wscore
99013     2008   1460      43
126840    2013   1366      43
126982    2013   1442      43
15619     1989   1343      43
122700    2012   1349      43


**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [48]:
# Assuming the dataframe sorted by 'Wscore' in descending order is 'df_wscore_sorted'
highest_winning_score_game = df_wscore_sorted.iloc[-1]
season = highest_winning_score_game['Season']
wteam = highest_winning_score_game['Wteam']
print(f"The game with the highest winning score is in season {season} by team {wteam}.")

The game with the highest winning score is in season 2013 by team 1380.


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [28]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [29]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [50]:

# Re-importing numpy and pandas as the previous session was reset
import numpy as np
import pandas as pd

#then we take the difference between Wscore and Lscore 
df['ScoreDifference'] = (df['Wscore'] - df['Lscore']).abs()

average_score_difference = df['ScoreDifference'].mean()

filtered_df = df[df['ScoreDifference'] > average_score_difference]

print(filtered_df)

# We can now see the variables 


        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
1

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [51]:
# Sorting the filtered dataframe by 'ScoreDifference' in descending order to find the game with the highest 'ScoreDifference'
# Extracting the row for the game with the highest 'ScoreDifference'
highest_difference_game = filtered_df.loc[filtered_df['ScoreDifference'].idxmax()]

## Extracting the season, winning team, and losing team for the game with the highest 'ScoreDifference'
season = highest_difference_game['Season']
winning_team = highest_difference_game['Wteam']
losing_team = highest_difference_game['Lteam']

print(f"The game with the highest score difference occurred in season {season}, "
      f"between team {winning_team} and team {losing_team}.")

#now we know the teams who had a high score difference 

The game with the highest score difference occurred in season 1996, between team 1409 and team 1341.


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [32]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [33]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       N         3
       A         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [34]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [35]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [36]:
# Grouping the DataFrame by 'season' and calculating the average winning score for each season

average_winning_score_per_season = df.groupby('Season')['Wscore'].mean()
print(average_winning_score_per_season)

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


In [ ]:
#the higehest being 2016

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [37]:
# Grouping the DataFrame by 'Wteam' and finding the maximum winning score for each team across all seasons
max_winning_score_per_team = df.groupby('Wteam')['Wscore'].max()

print(max_winning_score_per_team)

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [38]:
# Grouping the DataFrame by both 'season' and 'Wteam', and calculating the average winning score for each group
avg_winning_score_per_season_team = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

# Finding the team with the highest average winning score for each season
highest_avg_winning_score_per_season_team = avg_winning_score_per_season_team.groupby('Season').idxmax()

result_teams = avg_winning_score_per_season_team.loc[highest_avg_winning_score_per_season_team].reset_index()
print(result_teams)

    Season  Wteam      Wscore
0     1985   1328   92.800000
1     1986   1109   91.200000
2     1987   1380   95.875000
3     1988   1258  111.750000
4     1989   1258  117.315789
5     1990   1258  126.347826
6     1991   1380  112.312500
7     1992   1380   99.642857
8     1993   1380  101.875000
9     1994   1380  106.583333
10    1995   1206  102.833333
11    1996   1206  103.600000
12    1997   1254   94.142857
13    1998   1395  101.153846
14    1999   1317   95.750000
15    2000   1395   95.647059
16    2001   1395   99.315789
17    2002   1242   92.888889
18    2003   1395   88.222222
19    2004   1194   92.142857
20    2005   1311   91.333333
21    2006   1190   89.533333
22    2007   1440   99.400000
23    2008   1440   94.500000
24    2009   1377   96.200000
25    2010   1440   95.000000
26    2011   1449   90.130435
27    2012   1370   87.400000
28    2013   1377   86.500000
29    2014   1322   92.600000
30    2015   1322   90.733333
31    2016   1146   97.142857


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [39]:
# Grouping the DataFrame by both 'season' and 'Wteam', and counting the number of wins for each team in each season
win_count_per_team_per_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='WinCount')

print(win_count_per_team_per_season)

       Season  Wteam  WinCount
0        1985   1102         5
1        1985   1103         9
2        1985   1104        21
3        1985   1106        10
4        1985   1108        19
...       ...    ...       ...
10167    2016   1460        20
10168    2016   1461        12
10169    2016   1462        27
10170    2016   1463        21
10171    2016   1464         9

[10172 rows x 3 columns]


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [40]:
# Finding the team with the most wins for each season, I'm gonna dived the variabels in three boxes, "season, Wteam and wincount"
win_count_per_team_per_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='WinCount')

team_with_most_wins_per_season = win_count_per_team_per_season.groupby('Season')['Wteam'].idxmax()

result_df = win_count_per_team_per_season.loc[team_with_most_wins_per_season]


print(result_df)

# now i can see the team with the most win, which is season 1998 with 18 wincount 

       Season  Wteam  WinCount
281      1985   1464        16
563      1986   1464        10
852      1987   1464         8
1142     1988   1464         6
1435     1989   1464         3
1727     1990   1464         6
2022     1991   1464         8
2319     1992   1464         3
2616     1993   1464         3
2917     1994   1464         5
3218     1995   1464        16
3522     1996   1464        10
3827     1997   1464         8
4133     1998   1464        18
4442     1999   1464        12
4758     2000   1464        11
5076     2001   1464        17
5397     2002   1464         3
5724     2003   1464         7
6050     2004   1464         7
6379     2005   1464         3
6713     2006   1464         6
7049     2007   1464        12
7390     2008   1464         7
7737     2009   1464         9
8084     2010   1464         6
8429     2011   1464         7
8774     2012   1464        14
9120     2013   1464        14
9471     2014   1464        12
9820     2015   1464         8
10171   

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [41]:
# Selects the first six columns of the df DataFrame and assigns it to df_1
df_1 = df.iloc[:,:6]

# Drops the 'Daynum', 'Wteam', and 'Wscore' columns from df_1

df_1 = df_1.drop(columns =['Daynum','Wteam','Wscore'])

#Groups df_1 by 'Season' and 'Lteam', calculates the mean for each group, and prints the result.
print(df_1.groupby(['Season','Lteam']).mean())

print(average_winning_score_per_season)

                 Lscore
Season Lteam           
1985   1102   61.000000
       1103   55.142857
       1104   60.111111
       1106   69.142857
       1108   74.000000
...                 ...
2016   1460   61.230769
       1461   68.277778
       1462   71.200000
       1463   61.333333
       1464   71.238095

[10183 rows x 1 columns]
Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [42]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [43]:
# This code iterates through each row in my DataFrame df, checks if the winning score is greater than 100 for that row, and appends either 'Yes' or 'No' to the high_scoring_game
df['HighScoringGame'] = df['Wscore'].apply(lambda x: 'Yes' if x > 100 else 'No')

print(df.head())

# In the provided sample, 'HighScoringGame' is 'No' for all rows, suggesting that none of the games in this subset had an exceptionally high winning score.

   Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   
3    1985      25   1165      70   1432      54    H      0               16   
4    1985      25   1192      86   1447      74    H      0               12   

  HighScoringGame  
0              No  
1              No  
2              No  
3              No  
4              No  


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [44]:
#this line calculates the number of wins for each team by counting the occurrences of each unique value in the 'Wteam' column
win_counts = df['Wteam'].value_counts()

#similarly, this line calculates the number of losses for each team by counting the occurrences of each unique value in the 'Lteam' column. 
loss_counts = df['Lteam'].value_counts()

#this line adds the counts of wins and losses for each team to calculate the total number of games played
total_games = win_counts.add(loss_counts, fill_value=0)

total_games_dict = total_games.to_dict()

print(total_games_dict)

{1101: 76, 1102: 840, 1103: 910, 1104: 975, 1105: 447, 1106: 855, 1107: 512, 1108: 866, 1109: 169, 1110: 910, 1111: 876, 1112: 981, 1113: 937, 1114: 910, 1115: 520, 1116: 980, 1117: 891, 1118: 27, 1119: 868, 1120: 936, 1121: 175, 1122: 903, 1123: 906, 1124: 906, 1125: 506, 1126: 906, 1127: 442, 1128: 106, 1129: 907, 1130: 951, 1131: 945, 1132: 895, 1133: 949, 1134: 200, 1135: 847, 1136: 240, 1137: 912, 1138: 728, 1139: 923, 1140: 965, 1141: 872, 1142: 597, 1143: 947, 1144: 850, 1145: 934, 1146: 267, 1147: 698, 1148: 844, 1149: 824, 1150: 942, 1151: 899, 1152: 865, 1153: 970, 1154: 820, 1155: 950, 1156: 900, 1157: 784, 1158: 720, 1159: 887, 1160: 928, 1161: 905, 1162: 831, 1163: 977, 1164: 893, 1165: 833, 1166: 957, 1167: 260, 1168: 891, 1169: 724, 1170: 660, 1171: 829, 1172: 879, 1173: 960, 1174: 913, 1175: 890, 1176: 526, 1177: 942, 1178: 920, 1179: 914, 1180: 930, 1181: 1013, 1182: 924, 1183: 871, 1184: 863, 1185: 900, 1186: 849, 1187: 861, 1188: 207, 1189: 481, 1190: 891, 1191: 928,

In [ ]:
#the result is a dictionary where each team ID is associated with the total number of games they played. This provides a summary of each team's participation in the dataset.

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [45]:
#this line calculates the absolute score difference for each game and adds it as a new column 
df['ScoreDifference'] = (df['Wscore'] - df['Lscore']).abs()

#this line groups the dataframe df by the season column and finds the index of the row with the maximum score difference within each season.
idx = df.groupby('Season')['ScoreDifference'].idxmax()

highest_score_diff_per_season = df.loc[idx]

print(highest_score_diff_per_season)

#the result highest_score_diff_per_season contains the rows from theoriginal dataframe where the score difference was highest for each season this allows me to identify the games with the most score differences in each season.

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410    

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [46]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [47]:
#this line calculates the number of wins for each team by counting the occurrences of each unique value
win_counts = df['Wteam'].value_counts()

#same as before this line calculates the number of losses for each team by counting the occurrences of each unique value
loss_counts = df['Lteam'].value_counts()

#and here it adds the counts of wins and losses for each team to calculate the total number of games played.
total_games = win_counts.add(loss_counts, fill_value=0)

print(total_games)

1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Name: count, Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [48]:
#here it calculates the score difference for each game by subtracting the losing score from the winning score 
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

#this line groups the dataframe df by the season column and finds the index of the row with the maximum scoredifference within each season.
idx = df.groupby('Season')['ScoreDifference'].idxmax()

#using the indices found in the previous step, this line selects the rows with the highest score difference for each season

games_with_highest_score_diff = df.loc[idx]

print(games_with_highest_score_diff)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
236       1985      33   1361     128   1288      68    H      0   
4731      1986      60   1314     129   1264      45    N      0   
8240      1987      51   1155     112   1118      39    H      0   
12046     1988      40   1328     152   1147      84    H      0   
16677     1989      64   1242     115   1135      45    H      0   
19502     1990      26   1181     130   1217      54    H      0   
25161     1991      73   1163     115   1148      47    H      0   
27997     1992      30   1116     128   1126      46    H      0   
33858     1993      86   1328     146   1197      65    H      0   
36404     1994      47   1228     121   1152      52    H      0   
39858     1995      26   1246     124   1404      50    H      0   
44653     1996      48   1409     141   1341      50    H      0   
49033     1997      56   1278     114   1106      34    H      0   
52600     1998      33   1395     153   1410    

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [49]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [50]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [51]:
type(df['Wscore'])

pandas.core.series.Series

In [52]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [53]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [54]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [55]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y